### ISSUE TITLE
Stale _astroActionPayload cookie breaks all SSR pages

### ISSUE THREAD

**Issue by mlafeldt** [withastro/astro#12063]
### Astro Info

```block
Astro                    v4.15.9
Node                     v22.8.0
System                   macOS (arm64)
Package Manager          bun
Output                   hybrid
Adapter                  none
Integrations             none
```

### Describe the Bug

~~As the title says,~~ returning `NOT_FOUND` from actions breaks SSR when you have a static 404 page.

This happens because the `_astroActionPayload` cookie never gets deleted in that case (I guess because there's no server). As a result, _all_ SSR routes will return the 404 page until you manually clear that cookie.

See https://github.com/mlafeldt/extraterrestrial-equator for a small repo.

### What's the expected result?

Reloading SSR routes should work even with a static 404 page.

### Link to Minimal Reproducible Example

https://github.com/mlafeldt/extraterrestrial-equator

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by mlafeldt** [comment]
Another interesting fact: With the Cloudflare adapter, server-rendering `404.astro` via `export const prerender = false` will *not* fix the problem. However, doing the same with the default Node adapter will fix it (see repro). 🤔

**Comment by mlafeldt** [comment]
This can happen with other status codes too.

One of my apps is stuck/broken with a 500 error because `_astroActionPayload` contains this after one Astro action returned a 500:

```
{"actionName":"user.deleteUser","actionResult":{"type":"error","status":500,"contentType":"application/json","body":"{\"type\":\"AstroActionError\",\"code\":\"INTERNAL_SERVER_ERROR\",\"status\":500,\"message\":\"User not found\"}"}}
```

**Comment by mlafeldt** [comment]
In fact, you can break all SSR routes of an Astro site by doing this:

```
document.cookie = "_astroActionPayload=boom"
```

This will even affect routes that don't use Astro Actions at all, which shows a possible unintended consequence of the current design.

**Comment by michaelpayne02** [comment]
I encountered this when creating a repro for #11675 (and #11973). This error also occurs if the action call fails schema validation regardless if the handler returns an error code or not.

In 14.5.9 this is the stack trace

```00:50:13 [ERROR] Unexpected token 'd', "deleted" is not valid JSON
  Stack trace:
    at JSON.parse (<anonymous>)
    at eval (/home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/actions/runtime/middleware.js:27:103)
    at eval (/home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/core/middleware/sequence.js:21:12)
    at eval (file:///home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/core/middleware/sequence.js:58:18)
    at applyHandle (file:///home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/core/middleware/sequence.js:33:22)
   ```

After #12016 the result is base64-encoded for size savings and parsing still throws
```
01:07:00 [ERROR] Invalid character
  Stack trace:
    at decodeBase64_internal (file:///home/projects/github-g63nyt-zhjdjc/node_modules/@oslojs/encoding/dist/base64.js:90:23)
    at eval (/home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/actions/runtime/middleware.js:33:75)
    at eval (/home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/core/middleware/sequence.js:21:12)
    at eval (file:///home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/core/middleware/sequence.js:58:18)
    at applyHandle (file:///home/projects/github-g63nyt-zhjdjc/node_modules/astro/dist/core/middleware/sequence.js:33:22)
```
So two things:
1. Somehow the cookie is not actually being removed on the server when set to `deleted` in the middleware chain and is leaking to the client.
2. Because the parsing errors are server-side, we should be able to redirect to the correct page using the Referrer header with the payload cleared

**Comment by mlafeldt** [comment]
In general, I wonder how using a single cookie for all actions is supposed to work when, for example, two or more actions happen simultaneously.

**Comment by bholmesdev** [comment]
Fair point @mlafeldt! However, it's worth noting cookies are only used for server form requests. It should only be possible to handle and render a single form request at a time using the standard form request model. If you submit multiple actions at once via client-side JS, everything works as expected.

As for the issue at hand: we're considering a model that would make cookie forwarding an opt-in at the action level. This work may negate the bug entirely, but we'll be sure to capture in an integration test.

**Comment by jamesknelson** [comment]
A possibly related issue: middleware redirects can cause an action's cookie to not be deleted, and in turn to prevent future actions from being processed.

The particular steps required in my case:

- I have a `logout` action, which is available in every page via a button in the layout
- I also have a middleware that redirects logged out users to a login screen:

```ts
      const isAuthenticationRequired =
        !context.url.pathname.startsWith('/lobby');
      if (isAuthenticationRequired) {
        if (!context.locals.identity) {
          return context.redirect(`/lobby/login`);
        }
      }
```

The redirect middleware and the login screen worked fine until I introduced the logout button. From what I can tell, the reason is that the redirect middleware runs in the same request as the logout action is processed, redirecting to the login screen with the logout button's action cookie still in place, which somehow prevents login actions from being submitted at all.

Here's the stale cookie I get after decoding, which stays in place after reloads/form submissions. The only way to get the app to start responding again is to manually deleted it.

```
{"actionName":"logout","actionResult":{"type":"data","status":200,"contentType":"application/json+devalue","body":"[true]"}}```

FWIW, I got around this by handling `logout` as an endpoint function instead of an action.

I've noticed that you don't have a `redirect` function available on `ActionAPIContext`, and wonder if you don't want people redirecting to prevent precisely this behavior?

**Comment by bholmesdev** [comment]
Alright, we spent some time reevaluating how form actions should be handled. We landed on a strategy that we'll be shipping in the next 5.0 Beta:

- **Cookie redirects are no longer built-in.** Due to the 4 KB size limit and nuanced behavior for hybrid apps, we don't think it's a best practice we should endorse. This means that action results will render as a standard form POST without redirects. Confirmed the issue reported here will go away using this new default.
- If you _do_ want to implement cookie redirects, or build your own behavior, **you will have access to action results from middleware.** We will add a new `getActionContext()` utility to check for inbound action requests and handle results however you wish. Store in your own session storage, store as a cookie, add your own gated security layer, etc. [Currently in PR.](https://github.com/withastro/astro/pull/12373)

This should make the issue reported here addressable in user-land. For example, you could forward action results as a cookie for server-based routes, and avoid redirecting when a `NOT_FOUND` error is encountered. 

There's a few features I'd like to see Astro address in the future to make our defaults simpler:
- Built-in session storage to plug in your favorite storage provider. [Now in stage 2 discussion](https://github.com/withastro/roadmap/issues/1050)
- Some primitive to check whether a given route is pre-rendered or not. In the case of this bug, we should only redirect with a cookie when the destination route is server-rendered. Otherwise, the cookie cannot be deleted.

This will be shipped in Beta, so we're open to feedback! Thanks to @mlafeldt @jamesknelson and @michaelpayne02 for sharing examples with us.

**Comment by bholmesdev** [comment]
This change has now been shipped in Astro 5.0. You can [follow the migration guide from 4.X here](https://docs.astro.build/en/guides/upgrade-to/v5/#changed-actions-submitted-by-html-forms-no-longer-use-cookie-redirects). If you run into issues, please open a new issue. Thanks everyone!
